### ISSUE TITLE
Avoid large graphs through custom MutableMappings

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#1763]
For very large computations, such as sometimes occur in Dask.array, we sometimes run into a limit of graph size.  This might occur, for example, when we have 100's of thousands of chunks.  

On the scheduler side I suspect that this limitation will continue into the medium-term future at least  However there might be some things we can do on the client side to avoid frequently creating and copying many large arrays.  Both of these involve relaxing the type of the `.dask` attribute from `dict` to `MutableMapping` and constructing clever mutable mappings that are more efficient in our case.

1.  We could construct a MutableMapping that is a union of several dictionaries.  This is roughly what we do with dask.delayed to avoid overzealous copying there.  Making a "copy"  generally consists of making a shallow copy of a list of dicts.  Updating is just appending to this list.  Many MutableMapping operations become slower than constant time, but we rarely care about these operations.
2.  We could build a programmatic MutableMapping.  Some of our dicts are fairly simple dict comprehensions.  We could encode these more efficiently with some programmatic structure.  For dask.array particularly some large effort could also go into special dicts for atop, which would be difficult but also cover a large range.

This would help avoid copies.  It would possibly help to reduce client-scheduler communication.  It would not help to reduce scheduling overhead.

**Comment by jcrist** [comment]
New in Python 3.3, but this could be useful: https://docs.python.org/3.5/library/collections.html#collections.ChainMap.

**Comment by crusaderky** [comment]
This looks like a very major effort to me but probably worth it.

I've been playing with ChainMap in dask for a while now. The feedback is generically positive as long as you understand a few caveats:

- seemingly effortless functions, like \_\_len\_\_, actually build a set of keys on the fly so they're incredibly slow and you must be super careful to avoid them when coding (e.g. not use \_\_len\_\_ as a "is empty" test). Other bad cases are e.g. dict(mymap) to convert a ChainMap to a dict, or ChainMap.items() or ChainMap.values(). These functions are not optimized, as they will iterate on all keys, create a set of seen keys, and then retrieve the values by sequentially accessing all maps in the list with a O(n^2) computation time.
- more in general, if run any function on them that actually needs to retrieve data, like everything in dask.optimizer or like the auto-cull on Array.\_\_getitem\_\_, all your unit tests will work and you'll probably not notice how bad performance is until you run your code with production-scale data.

Overall I would not recommend using ChainMap directly in dask. Probably a subclass that overrides some methods with a variant more tailored to dask would be better, e.g.
```
def update(self, rhs):
    if isinstance(rhs, ChainMap):
        self.maps += [m for m in rhs.maps if len(m) != 0]
    else:
        self.maps.append(rhs)
````

So I'm trying to come up with a list of things that need to be done to migrate dask from a dict backend to a ChainMap or similar object - please help me with anything I might have missed:

- change all init methods and transformation functions
- write unit tests for all transformation functions to verify that in the end your dask attribute is a flat ChainMap with multiple elements in its maps list
- update Array.\_\_init\_\_ as well as a @property setter that converts the dask attribute from dict to ChainMap for the purpose of backwards compatibility with custom user code
- update all optimizer and get() functions, also in dask.distributed, to accept either a plain dict or a ChainMap; in the latter case the ChainMap must be converted to dict before proceeding to avoid abysmal performance
- contribute to Python 3.7 to implement optimized functions for ChainMap

What else?

**Comment by mrocklin** [comment]
Hi @crusaderky thanks for looking into this.  A few thoughts.

1.  I agree that if we were to do this we would probably want to rewrite a number of dask.array functions to be more amenable to this style
2.  Given that we need to support Python 2, we might consider just implementing our own ChainMap with the routines that we need.  I don't expect this to be particularly hard.
3.  Testing all dask.array operations is actually pretty easy because we use our own [assert_eq](https://github.com/dask/dask/blob/master/dask/array/utils.py#L33-L69) function.  A couple of added checks to this function should effectively test all dask.array operations.
4.  Coercing all MutableMappings to dict prior to optimization or get seems reasonable.
5.  Rather than think about `update` we might want to think about the pure variant, `merge`, which tends to be more often used throughout dask and is nice because it's a bare function (and so easier to re-implement on our own) than a method call.

**Comment by crusaderky** [comment]
The biggest problem of a ChainMap is that it risks making life harder for projects with diamond dependencies.

e.g.``y = f1(x) + f2(x) + f3(x) + f4(x)``
Right now the dask dict of x appears only once in y. If the backend moved to ChainMap or equivalent, you'd find yourself with 4 copies. _normally_ they should be actually 4 references to the same dict, but that's quite flimsy and easy to break.

A possibly healthier solution may be to redesign the dask dict as follows:
```
y.dask = {
    y.name: {<key/values added while summing up the outputs of f1, f2, f3, f4>},
    f1.name: {<key/values added while creating f1>},
    f2.name: {<key/values added while creating f2>},
    f3.name: {<key/values added while creating f3>},
    f4.name: {<key/values added while creating f4>},
    x.name: {<key/values added while creating x>},
}
```
By default, this would translate to 1 dict per dask transformation, but there should be a system for higher-level libraries to map it to 1 dict per higher-level operation, e.g. xarray could store in the same dict all the transpose and broadcast-related keys; an even higher level library (like the one I work on) would collapse f1, f2, f3 and f4 into the dict of y.

**Comment by mrocklin** [comment]
You might want to check out the implementation in delayed, which already
does something similar.  It keeps a list around and only adds dicts if
their object id isn't already in the list.  Keying by name sounds like a
nice idea though.

**Comment by mrocklin** [comment]
Thinking about this a bit more, given that we probably *have* to expand out each task once, at least at the end, thinking about deduplication of component dictionaries may be more valuable than thinking about programmatic dictionaries.

**Comment by jcrist** [comment]
I believe this was for the most part fixed in #1985. Closing.
